apps/aecore/src/aec_block_generator.erl

```diff
@@ -0,0 +1,272 @@
+        , candidate = undefined :: undefined | aec_blocks:block()
+        , candidate_state = undefined :: undefined | term()
+        , new_txs = []          :: list(term())
+        }).
+
+-include("common.hrl").
+-include("blocks.hrl").
+
+%% -- API --------------------------------------------------------------------
+start_link() ->
+    gen_server:start_link({local, ?MODULE}, ?MODULE, [], []).
+
+stop() ->
+    gen_server:stop(?MODULE).
+
+-spec get_candidate() -> {ok, aec_blocks:block()} | {error, no_candidate}.
+get_candidate() ->
+    try
+        gen_server:call(?MODULE, get_candidate, 100)
+    catch _:_ ->
+        {error, no_candidate}
+    end.
+
+start_generation() ->
+    gen_server:cast(?MODULE, start_generation).
+
+stop_generation() ->
+    gen_server:cast(?MODULE, stop_generation).
+
+%% -- gen_server callbacks ---------------------------------------------------
+
+init([]) ->
+    aec_events:subscribe(tx_created),
+    aec_events:subscribe(tx_received),
+    aec_events:subscribe(block_created),
+    aec_events:subscribe(top_changed),
+    {ok, #state{}}.
+
+handle_call(get_candidate, _From, State = #state{ candidate = undefined }) ->
+    {reply, {error, no_candidate}, State};
+handle_call(get_candidate, _From, State = #state{ candidate = Candidate }) ->
+    {reply, {ok, Candidate}, State};
+handle_call(Req, _From, State) ->
+    lager:info("Unexpected call: ~p", [Req]),
+    {reply, ok, State}.
+
+handle_cast(start_generation, State) ->
+    lager:debug("start_generation"),
+    {noreply, do_start_generation(State)};
+handle_cast(stop_generation, State) ->
+    lager:debug("stop_generation"),
+    {noreply, do_stop_generation(State)};
+handle_cast({new_candidate, Candidate, CandidateState}, State) ->
+    epoch_mining:info("New candidate generated", []),
+    lager:debug("New candidate generated", []),
+    publish_candidate(Candidate),
+    State1 = finish_worker(State),
+    State2 = State1#state{ candidate = Candidate
+                         , candidate_state = CandidateState },
+    {noreply, maybe_start_worker_txs(State2)};
+handle_cast({worker_done, Result}, State = #state{ worker = {Pid, _}}) ->
+    State1 = finish_worker(State),
+    lager:debug("Candidate worker ~p done ~p", [Pid, Result]),
+    {noreply, State1};
+handle_cast(Msg, State) ->
+    lager:info("Unexpected message ~p", [Msg]),
+    {noreply, State}.
+
+handle_info({gproc_ps_event, Event, _Data}, State = #state{ generating = false }) ->
+    lager:debug("ignored event ~p", [Event]),
+    {noreply, State};
+handle_info({gproc_ps_event, Event, #{info := Info}}, State) ->
+    lager:debug("got event ~p", [Event]),
+    State1 =
+        case Event of
+            block_created -> preempt_generation(State, Info);
+            top_changed   -> preempt_generation(State, Info);
+            tx_created    -> add_new_tx(State, Info);
+            tx_received   -> add_new_tx(State, Info);
+            _             -> State
```

Comment:
As the process subscribes only for certain events, I expect this clause not to be exercised unless of typos. Maybe logging a warning here? (Totally optional.)